### PR TITLE
Add unit tests for k8sd/setup/kube_apiserver

### PR DESCRIPTION
**Pull request by eaudetcobello**
Needs https://github.com/canonical/k8s-snap/pull/201

### CODE REVIEW

File: src/k8s/pkg/k8sd/setup/kube_apiserver.go [hunk 1]
```diff
@@ -80,7 +80,7 @@ func KubeAPIServer(snap snap.Snap, serviceCIDR string, authWebhookURL string, en
 	case "k8s-dqlite":
 		args["--etcd-servers"] = fmt.Sprintf("unix://%s", path.Join(snap.K8sDqliteStateDir(), "k8s-dqlite.sock"))
 	default:
-		return fmt.Errorf("unsupported datastore %s. must be 'k8s-dqlite'", datastore)
+		return fmt.Errorf("unsupported datastore '%s'. must be 'k8s-dqlite'", datastore)
```
Comment (bschimke95) [inline src/k8s/pkg/k8sd/setup/kube_apiserver.go:83]: you can use `%q` for that

File: src/k8s/pkg/k8sd/setup/kube_apiserver_test.go [hunk 2]
```diff
@@ -28,5 +33,123 @@ func TestKubeAPIServer(t *testing.T) {
...
+			})
+		}
+
+		// Ensure the kube-apiserver arguments file has exactly the expected number of arguments
+		args, err := utils.ParseArgumentFile(path.Join(s.Mock.ServiceArgumentsDir, "kube-apiserver"))
+		g.Expect(err).To(BeNil())
+		g.Expect(len(args)).To(Equal(len(tests)))
+	})
+
+	t.Run("Configure kube-apiserver with without proxy", func(t *testing.T) {
```
Comment (bschimke95) [inline src/k8s/pkg/k8sd/setup/kube_apiserver_test.go:96]: ~with~

File: src/k8s/pkg/k8sd/setup/kube_apiserver_test.go [hunk 3]
```diff
@@ -28,5 +33,123 @@ func TestKubeAPIServer(t *testing.T) {
...
+	t.Run("Handle unsupported datastore", func(t *testing.T) {
+		g := NewWithT(t)
+
+		// Create a mock snap
+		s, _ := testApiServerFixture(t)
+
+		// Attempt to configure kube-apiserver with an unsupported datastore
+		err := setup.KubeAPIServer(s, "10.0.0.0/24", "https://auth-webhook.url", false, "unsupported-datastore", "Node,RBAC")
+		g.Expect(err).ToNot(BeNil())
+		g.Expect(err.Error()).To(ContainSubstring("unsupported datastore"))
```
Comment (bschimke95) [inline src/k8s/pkg/k8sd/setup/kube_apiserver_test.go:153]: TIL: https://github.com/onsi/gomega/blob/master/matchers.go#L114

`g.Expect(err).To(MatchError(ContainsSubstring("unsupported datastore"))`

**Review by bschimke95 — COMMENTED** [review]
Looks good - added some comments. Also, comments from the other PR apply here (e.g. about the fixture naming)

File: src/k8s/pkg/k8sd/setup/kube_apiserver.go [hunk 4]
```diff
@@ -80,7 +80,7 @@ func KubeAPIServer(snap snap.Snap, serviceCIDR string, authWebhookURL string, en
 	case "k8s-dqlite":
 		args["--etcd-servers"] = fmt.Sprintf("unix://%s", path.Join(snap.K8sDqliteStateDir(), "k8s-dqlite.sock"))
 	default:
-		return fmt.Errorf("unsupported datastore %s. must be 'k8s-dqlite'", datastore)
+		return fmt.Errorf("unsupported datastore %q. must be \"k8s-dqlite\"", datastore)
 	}
```
Comment (neoaggelos) [inline src/k8s/pkg/k8sd/setup/kube_apiserver.go:84]: not related to the PR, please revert

**Review by neoaggelos — COMMENTED** [review]
Final comment, and this is good to go

**Review by neoaggelos — APPROVED** [review]

**Comment by eaudetcobello** [comment]
@neoaggelos Added a test for external datastore url, can you check and comment/reapprove?

**Review by neoaggelos — APPROVED** [review]

**Pull request merged**